src/streams.h

```diff
@@ -13,36 +13,56 @@
 
 #include <algorithm>
 #include <assert.h>
+#include <crypto/common.h>
 #include <cstddef>
 #include <cstdio>
+#include <cstring>
 #include <ios>
 #include <limits>
 #include <optional>
 #include <stdint.h>
 #include <string.h>
 #include <string>
-#include <utility>
 #include <vector>
+#include <util/check.h>
 
 namespace util {
-inline void Xor(Span<std::byte> write, Span<const std::byte> key, size_t key_offset = 0)
+inline void XorInt(Span<std::byte> write, const uint64_t key, const size_t size)
 {
-    if (key.size() == 0) {
-        return;
+    Assume(key);
+    assert(size <= write.size());
+    uint64_t raw = 0;
+    memcpy(&raw, write.data(), size);
+    raw ^= key;
+    memcpy(write.data(), &raw, size);
+}
+inline void Xor(Span<std::byte> write, const uint64_t key)
+{
+    Assume(key);
+    for (constexpr auto size{sizeof(key)}; write.size() >= size; write = write.subspan(size)) {
+        XorInt(write, key, size);
     }
-    key_offset %= key.size();
-
-    for (size_t i = 0, j = key_offset; i != write.size(); i++) {
-        write[i] ^= key[j++];
-
-        // This potentially acts on very many bytes of data, so it's
-        // important that we calculate `j`, i.e. the `key` index in this
-        // way instead of doing a %, which would effectively be a division
-        // for each byte Xor'd -- much slower than need be.
-        if (j == key.size())
-            j = 0;
+    switch (write.size()) { // Help the compiler specialize 1, 2 and 4 byte cases
+    case 0:  break;
+    case 1:  XorInt(write, key, 1); break;
+    case 2:  XorInt(write, key, 2); break;
+    case 4:  XorInt(write, key, 4); break;
+    default: XorInt(write, key, write.size());
     }
 }
+
+inline uint64_t RotateKey(const uint64_t key, const size_t key_offset)
+{
+    Assume(key);
+    size_t key_rotation{8 * key_offset};
+    if (key_rotation % 64 == 0) return key;
```

Comment:
yes, it's an optimization to avoid doing any rotation when it would wrap around - it would work without this as well.
It's not a measurable speedup, though, so I can remove it if you insist.